### PR TITLE
Add Kuttl (declarative E2E test) implem

### PR DESCRIPTION
**Pull request by PERES-Richard**
| Q               | A
| --------------- | ---
| New feature    | [x]

### What's in this PR?
A functional way to test Casskop operator with end to end declarative tests by using Kuttl test framework.

### Why?
The more it's tested is the better it is. Also It could help during Cass-operator features integration.

### Checklist

- [x] Implementation tested
- [x] Logging code meets the guideline
- [x] User guide and development docs updated (if needed)
- [ ] Append changelog

### To Do
- [x] Update Doc (developer guide)
- [x] Add it to circle ci workflow
- [ ] Use new helm fix (see #282 ) for CRDs

### CODE REVIEW

File: test/e2e/kuttl/kuttl-test.yaml [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+timeout: 300
+commands:
+  - command: helm repo add orange-incubator https://orange-kubernetes-charts-incubator.storage.googleapis.com/
+    ignoreFailure: true
+  - command: kubectl apply -f https://raw.githubusercontent.com/Orange-OpenSource/casskop/master/deploy/crds/db.orange.com_cassandraclusters_crd.yaml
+    ignoreFailure: true
+  - command: kubectl apply -f https://raw.githubusercontent.com/Orange-OpenSource/casskop/master/deploy/crds/db.orange.com_cassandrabackups_crd.yaml
+    ignoreFailure: true
+  - command: kubectl apply -f https://raw.githubusercontent.com/Orange-OpenSource/casskop/master/deploy/crds/db.orange.com_cassandrarestores_crd.yaml
+    ignoreFailure: true
```
Comment (PERES-Richard) [inline test/e2e/kuttl/kuttl-test.yaml:15]: This need can be deleted if we merge #282.

File: test/e2e/kuttl/ScaleUpAndDownDC/00-assert.yaml [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    dc1-rack1:
+      cassandraLastAction:
+        name: Initializing
+        status: Done
+      phase: Running
+  lastClusterAction: Initializing
+  lastClusterActionStatus: Done
+  phase: Running
+#  seedlist:
+#  - cassandra-e2e-dc1-rack1-0.cassandra-e2e.kuttl-test
```
Comment (cscetbon) [inline test/e2e/kuttl/ScaleUpAndDownDC/00-assert.yaml:24]: if not used, let's remove them

File: test/e2e/kuttl/ScaleUpAndDownDC/00-createCluster.yaml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  name: cassandra-e2e
+spec:
+  nodesPerRacks: 1
+  cassandraImage: cassandra:latest
+  dataCapacity: "1Gi"
+  hardAntiAffinity: false
+  deletePVC: false
+  autoPilot: true
+  autoUpdateSeedList: false
+  bootstrapImage: orangeopensource/cassandra-bootstrap:0.1.7
```
Comment (cscetbon) [inline test/e2e/kuttl/ScaleUpAndDownDC/00-createCluster.yaml:19]: why do you need to specify it for the tests ?

File: test/e2e/kuttl/ScaleUpAndDownDC/01-scaleUp-dc.yaml [hunk 4]
```diff
@@ -0,0 +1,12 @@
+apiVersion: db.orange.com/v1alpha1
+kind: CassandraCluster
+metadata:
+  name: cassandra-e2e
+spec:
+  nodesPerRacks: 1
+  topology:
```
Comment (cscetbon) [inline test/e2e/kuttl/ScaleUpAndDownDC/01-scaleUp-dc.yaml:7]: you can just update the one at line 6 and remove the topology here

File: test/e2e/kuttl/ScaleUpAndDownDC/01-assert.yaml [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  currentReplicas: 2
+  replicas: 2
+---
+apiVersion: db.orange.com/v1alpha1
+kind: CassandraCluster
+metadata:
+  name: cassandra-e2e
+spec:
+  nodesPerRacks: 1
+  topology:
```
Comment (cscetbon) [inline test/e2e/kuttl/ScaleUpAndDownDC/01-assert.yaml:15]: No need to test the topology if you update the one at the root level

File: test/e2e/kuttl/ScaleUpAndDownDC/02-assert.yaml [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+  currentReplicas: 1
+  replicas: 1
+---
+apiVersion: db.orange.com/v1alpha1
+kind: CassandraCluster
+metadata:
+  name: cassandra-e2e
+spec:
+  nodesPerRacks: 1
+  topology:
```
Comment (cscetbon) [inline test/e2e/kuttl/ScaleUpAndDownDC/02-assert.yaml:19]: no need to test the topology like the comment before

File: test/e2e/kuttl/ScaleUpAndDownDC/02-scaleDown-dc.yaml [hunk 7]
```diff
@@ -0,0 +1,12 @@
+apiVersion: db.orange.com/v1alpha1
+kind: CassandraCluster
+metadata:
+  name: cassandra-e2e
+spec:
+  nodesPerRacks: 1
```
Comment (cscetbon) [inline test/e2e/kuttl/ScaleUpAndDownDC/02-scaleDown-dc.yaml:6]: just update that one 👆 and remove the topology

**Review by cscetbon — CHANGES_REQUESTED** [review]

File: test/e2e/kuttl/ScaleUpAndDownDC/00-createCluster.yaml (hunk 3 above)
Comment (PERES-Richard) [reply test/e2e/kuttl/ScaleUpAndDownDC/00-createCluster.yaml:19]: I ran into a version asssertion probleme with this image during my local tests (nothing asked and version 0.1.5 was used wich created assertion error). I think we can delete it.

File: test/e2e/kuttl/ScaleUpAndDownDC/01-scaleUp-dc.yaml (hunk 4 above)
Comment (PERES-Richard) [reply test/e2e/kuttl/ScaleUpAndDownDC/01-scaleUp-dc.yaml:7]: I think we can keep it this way but add more steps like : setup a 2 dc topo, scale 1 up at dc level et assert only this one is correctly scaleup, then scaleup at root level which should scaleup only the second dc and then scale down both with assertions. What do you think if I create such a flow ?

File: test/e2e/kuttl/ScaleUpAndDownDC/00-assert.yaml (hunk 2 above)
Comment (PERES-Richard) [reply test/e2e/kuttl/ScaleUpAndDownDC/00-assert.yaml]: Done

File: test/e2e/kuttl/ScaleUpAndDownDC/00-createCluster.yaml (hunk 3 above)
Comment (PERES-Richard) [reply test/e2e/kuttl/ScaleUpAndDownDC/00-createCluster.yaml]: Option deleted

File: test/e2e/kuttl/ScaleUpAndDownDC/01-scaleUp-dc.yaml (hunk 4 above)
Comment (PERES-Richard) [reply test/e2e/kuttl/ScaleUpAndDownDC/01-scaleUp-dc.yaml]: Ok lets focus on basic scaleup for this PR for now. Changes applied !

File: test/e2e/kuttl/ScaleUpAndDownDC/01-assert.yaml (hunk 5 above)
Comment (PERES-Richard) [reply test/e2e/kuttl/ScaleUpAndDownDC/01-assert.yaml]: Removed

File: test/e2e/kuttl/ScaleUpAndDownDC/02-assert.yaml (hunk 6 above)
Comment (PERES-Richard) [reply test/e2e/kuttl/ScaleUpAndDownDC/02-assert.yaml]: Removed aswell

File: test/e2e/kuttl/ScaleUpAndDownDC/02-scaleDown-dc.yaml (hunk 7 above)
Comment (PERES-Richard) [reply test/e2e/kuttl/ScaleUpAndDownDC/02-scaleDown-dc.yaml]: Updated :)

File: Makefile [hunk 8]
```diff
@@ -403,18 +403,28 @@ endif
...
 ifeq ($(E2E_ARGS),)
 	@echo "args are: ExecuteCleanup; RollingRestart ; ClusterScaleDown ; ClusterScaleUp ; ClusterScaleDownSimple" && exit 1
 endif
 	docker run --rm --network host --env GO111MODULE=on -v $(PWD):$(WORKDIR) -v $(KUBECONFIG):/root/.kube/config $(BUILD_IMAGE):$(OPERATOR_SDK_VERSION) /bin/bash -c 'operator-sdk test local ./test/e2e --debug --image $(E2EIMAGE) --go-test-flags "-v -timeout 60m -run ^TestCassandraCluster$$/^group$$/^$(E2E_ARGS)$$" --operator-namespace cassandra-e2e' || { kubectl get events --all-namespaces --sort-by .metadata.creationTimestamp ; exit 1; }
 
+kuttl-test-fix-arg:
+ifeq ($(KUTTL_ARGS),)
+	@echo "args are: ScaleUpAndDown" && exit 1
+endif
+	kuttl test --config ./test/e2e/kuttl/kuttl-test.yaml ./test/e2e/kuttl --test $(KUTTL_ARGS) || { kubectl get events --all-namespaces --sort-by .metadata.creationTimestamp ; exit 1; }
```
Comment (cscetbon) [inline Makefile:426]: I think kuttl already collects events

File: helm/cassandra-operator/crds/db.orange.com_cassandrabackups_crd.yaml [hunk 9]
```diff
@@ -0,0 +1,136 @@
+apiVersion: apiextensions.k8s.io/v1
```
Comment (cscetbon) [inline helm/cassandra-operator/crds/db.orange.com_cassandrabackups_crd.yaml:1]: why is helm changing here ? It should be updated in that PR. Let's have only the kuttl tests and nothing else

**Review by cscetbon — CHANGES_REQUESTED** [review]

File: helm/cassandra-operator/crds/db.orange.com_cassandrabackups_crd.yaml (hunk 9 above)
Comment (PERES-Richard) [reply helm/cassandra-operator/crds/db.orange.com_cassandrabackups_crd.yaml]: Reverted :)

**Review by cscetbon — APPROVED** [review]

**Review by cscetbon — APPROVED** [review]

**Pull request merged**